### PR TITLE
Add support for version 2 routing information

### PR DESCRIPTION
**Pull request by gregarndt**

### CODE REVIEW

**Comment by gregarndt** [comment]
@camd Could you take a look at this and let me know your thoughts.  There is a test for the expected behavior which is what I think we would want.

Basically prefer task.extra.treeherder.revision if its there, then v2 routes, and then v1 (which are the version-less routes)

**Comment by gregarndt** [comment]
Note to self: need to add a commit for updated shrinkwrap before deploying (addition of lodash and removal of lodash.merge)

File: src/treeherder/job_handler.js [hunk 1]
```diff
@@ -71,26 +71,59 @@ const EVENT_MAP = {
...
-  if (task.extra && task.extra.treeherder) {
-    let treeherder = task.extra.treeherder
-    if (treeherder.revision && treeherder.revision_hash) {
-      return [treeherder.revision, treeherder.revision_hash];
-    }
+  let revision = _.get(task, 'extra.treeherder.revision');
+  let revision_hash = _.get(task, 'extra.treeherder.revision_hash');
+
+  if (revision && revision_hash) {
+    return [revision, revision_hash];
```
Comment (camd) [inline src/treeherder/job_handler.js]: I'm wondering if we need to have both here.  If you only have the ``revision`` then you could still just return ``[rev, ""]``

File: src/treeherder/job_handler.js [hunk 2]
```diff
@@ -71,26 +71,59 @@ const EVENT_MAP = {
...
   }
 
-  // Find treeherder specific route
-  let route = task.routes.find((route) => {
+  // Find treeherder specific routes
+  let routes = task.routes.filter((route) => {
```
Comment (camd) [inline src/treeherder/job_handler.js]: So here's a gap in my knowledge.  the syntax of ``task.routes.filter((route) => {``
Is ``=>`` a shorthand for creating a function?

File: src/treeherder/job_handler.js (hunk 1 above)
Comment (gregarndt) [inline src/treeherder/job_handler.js]: Ah good point!

File: src/treeherder/job_handler.js (hunk 2 above)
Comment (gregarndt) [inline src/treeherder/job_handler.js]: so task.routes.filter takes a function, I'm passing in an anonymous function
```js
(route) => {
  return route.split('.')[0] === prefix;
}
```

Basically it iterates over an array, and passes that element to the function you provide, the function then returns true or false.  All elements that evaluated to true will be added to an array and returned, which is assigned to "routes" in this case.

What this is checking for are any routes that begin with the prefix mozilla-taskcluster is listening for, such as "tc-treeherder".  The reason for the change between "find" and "filter" in this PR is that during a transition a task might have both routes in it whereas "find" will only return the first element that evaluates true.

Thinking about this more though, maybe we wouldn't have both routes in a single task.  Maybe it's either the v1 route or the v2, but not both.  I suppose it doesn't hurt handling the case where both are in a single task though.

File: test/treeherder/parse_task_revision_test.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    let task = {
+      routes: [
+        `${prefix}.mozilla-inbound.abc`,
+        `${prefix}.v2.mozilla-inbound.def`,
+        'build.mozilla-inbound.ghi'
+      ],
+      extra: {
+        treeherder: {
+          revision: 'xyz',
+          revision_hash: 'tuv'
```
Comment (camd) [inline test/treeherder/parse_task_revision_test.js]: I guess you could add a test here where there's no ``revision_hash`` in task.extra

**Comment by camd** [comment]
Awesome.  Thanks for the explanation, that makes sense.  So r+, with my few comments.  Adjust as you see fit.  :)

File: test/treeherder/parse_task_revision_test.js (hunk 3 above)
Comment (gregarndt) [inline test/treeherder/parse_task_revision_test.js]: Yea I'll be changing these tests slightly based on your recommendation to only use task.extra.treeherder.revision and not care at all about the revision _hash

**Comment by camd** [comment]
LGTM: r+  :)

**Pull request merged**